modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/consistency/GridConsistencyGetWithRecoveryFuture.java

```diff
@@ -0,0 +1,205 @@
+ * Prepares recovery data.
+ * Records consistency violation event.
+ */
+public class GridConsistencyGetWithRecoveryFuture extends GridConsistencyAbstractGetFuture {
+    /** Context. */
+    GridCacheContext ctx;
+
+    /**
+     *
+     */
+    public GridConsistencyGetWithRecoveryFuture(
+        AffinityTopologyVersion topVer,
+        GridCacheContext ctx,
+        Collection<KeyCacheObject> keys,
+        boolean readThrough,
+        UUID subjId,
+        String taskName,
+        boolean deserializeBinary,
+        boolean recovery,
+        IgniteCacheExpiryPolicy expiryPlc,
+        boolean skipVals,
+        String txLbl,
+        MvccSnapshot mvccSnapshot) {
+        super(topVer,
+            ctx,
+            keys,
+            readThrough,
+            subjId,
+            taskName,
+            deserializeBinary,
+            recovery,
+            expiryPlc,
+            skipVals,
+            txLbl,
+            mvccSnapshot,
+            false);
+
+        this.ctx = ctx;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void onResult() {
+        if (isDone())
+            return;
+
+        if (checkIsDone())
+            onDone(checkAndFix());
+    }
+
+    /**
+     *
+     */
+    private boolean checkIsDone() {
+        for (IgniteInternalFuture fut : futs.values()) {
+            if (!fut.isDone())
+                return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns latest (by version) entry for each key with consistency violation.
+     */
+    private Map<KeyCacheObject, EntryGetResult> checkAndFix() {
+        Map<KeyCacheObject, EntryGetResult> newestMap = new HashMap<>();
+        Map<KeyCacheObject, EntryGetResult> fixedMap = new HashMap<>();
+
+        for (GridPartitionedGetFuture<KeyCacheObject, EntryGetResult> fut : futs.values()) {
+            for (Map.Entry<KeyCacheObject, EntryGetResult> entry : fut.result().entrySet()) {
+                KeyCacheObject key = entry.getKey();
+
+                EntryGetResult candidate = entry.getValue();
+
+                newestMap.putIfAbsent(key, candidate);
+
+                EntryGetResult newest = newestMap.get(key);
+
+                if (newest.version().compareTo(candidate.version()) < 0) {
+                    newestMap.put(key, candidate);
+                    fixedMap.put(key, candidate);
+                }
+
+                if (newest.version().compareTo(candidate.version()) > 0)
+                    fixedMap.put(key, newest);
+            }
+        }
+
+        recordConsistencyViolation(fixedMap);
+
+        return fixedMap;
+    }
+
+    /**
+     * @param fixedMap Fixed map.
+     */
+    private void recordConsistencyViolation(Map<KeyCacheObject, EntryGetResult> fixedMap) {
+        GridEventStorageManager evtMgr = ctx.gridEvents();
+
+        if (!evtMgr.isRecordable(EVT_CONSISTENCY_VIOLATION)) {
+            logger().warning("Consistency violation found but can not be recorded since " +
```

Comment:
It's not a good idea to print a key or value in case data is secured. 
For example PCI DSS requires to mask pans.
Entry version is useless in case key and value missed.

Reply:
Let's print entry version here.

You can understand the node with the inconsistency entry from version info.
If some node(or nodes) will appear more frequently you can make the conclusion about network or configuration issues with a certain node.